src/Access/LDAPAccessStorage.cpp

```diff
@@ -0,0 +1,265 @@
+#include <Access/LDAPAccessStorage.h>
+#include <Access/AccessControlManager.h>
+#include <Access/User.h>
+#include <Access/Role.h>
+#include <Common/Exception.h>
+#include <common/logger_useful.h>
+#include <ext/scope_guard.h>
+#include <Poco/Util/AbstractConfiguration.h>
+
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+}
+
+
+LDAPAccessStorage::LDAPAccessStorage(const String & storage_name_)
+    : IAccessStorage(storage_name_)
+{
+}
+
+
+void LDAPAccessStorage::setConfiguration(AccessControlManager * access_control_manager_, const Poco::Util::AbstractConfiguration & config, const String & prefix)
+{
+    // TODO: switch to passing config as a ConfigurationView and remove this extra prefix once a version of Poco with proper implementation is available.
+    const String prefix_str = (prefix.empty() ? "" : prefix + ".");
+
+    std::scoped_lock lock(mutex);
+
+    const bool has_server = config.has(prefix_str + "server");
+    const bool has_roles = config.has(prefix_str + "roles");
+
+    if (!has_server)
+        throw Exception("Missing 'server' field for LDAP user directory.", ErrorCodes::BAD_ARGUMENTS);
+
+    const auto ldap_server_cfg = config.getString(prefix_str + "server");
+    if (ldap_server_cfg.empty())
+        throw Exception("Empty 'server' field for LDAP user directory.", ErrorCodes::BAD_ARGUMENTS);
+
+    std::set<String> roles_cfg;
+    if (has_roles)
+    {
+        Poco::Util::AbstractConfiguration::Keys role_names;
+        config.keys(prefix_str + "roles", role_names);
+
+        // Currently, we only extract names of roles from the section names and assign them directly and unconditionally.
+        roles_cfg.insert(role_names.begin(), role_names.end());
+    }
+
+    access_control_manager = access_control_manager_;
+    ldap_server = ldap_server_cfg;
+    default_role_names.swap(roles_cfg);
+    roles_of_interest.clear();
+    role_change_subscription = access_control_manager->subscribeForChanges<Role>(
+        [this] (const UUID & id, const AccessEntityPtr & entity)
+        {
+            return this->processRoleChange(id, entity);
+        }
+    );
+}
+
+
+bool LDAPAccessStorage::isConfiguredNoLock() const
+{
+    return !ldap_server.empty() &&/* !roles.empty() &&*/ access_control_manager;
```

Comment:
There is nothing wrong in having no roles. I know it's quite useless, but nothing wrong.

Reply:
Right, so here they are allowed. No roles doesn't mean the storage is not configured.

Reply:
Removed, not needed anymore.